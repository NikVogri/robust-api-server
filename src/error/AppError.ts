interface IAppError {
    message: string;
    statusCode: number;
    thrownError?: Error | unknown;
    payload?: Record<string, unknown>;
}

export class AppError extends Error {
    thrownError?: Error | unknown;
    statusCode: number;
    payload?: Record<string, unknown>;

    constructor(public data: IAppError) {
        super(data.message);

        this.thrownError = data.thrownError;
        this.statusCode = data.statusCode;
        this.payload = data.payload;
    }

    static toStringifiableObject(error: Error | unknown) {
        if (!(error instanceof Error)) return {};

        const err: Record<string, unknown> = {};

        Object.getOwnPropertyNames(error).forEach((prop) => {
            if ((error[prop as keyof Error] as unknown) instanceof Error) {
                err[prop] = AppError.toStringifiableObject(error);
            } else err[prop] = error[prop as keyof Error];
        });

        return err;
    }
}
