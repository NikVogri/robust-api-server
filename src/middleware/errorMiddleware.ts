import { NextFunction, Request, Response } from 'express';
import { AppError } from '../error/AppError';

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export const errorMiddleware = (err: Error, req: Request, res: Response, next: NextFunction) => {
    let message = `Something broke!`;
    let statusCode = 500;
    let additionalPayload = {};
    let errorCode: string | undefined;

    if (err instanceof AppError) {
        message = err.message;
        statusCode = err.statusCode;
        errorCode = err.errorCode;

        if (err.payload) {
            additionalPayload = err.payload;
        }
    }

    res.status(statusCode).send({ message, errorCode, ...additionalPayload });
};
