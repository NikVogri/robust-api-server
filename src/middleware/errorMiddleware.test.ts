import { describe, expect, it, vi } from 'vitest';
import { errorMiddleware } from './errorMiddleware';
import { AppError } from '../error/AppError';

import { getRequest, getResponse } from '../../vitest.mock';

describe('Test errorMiddleware', () => {
    it('should send formatted response for unexpected error', async () => {
        const request = getRequest();
        const response = getResponse();
        const nextFn = vi.fn();

        const error = new Error('Some unknown error occured');

        vi.spyOn(response, 'status').mockImplementationOnce((...args) => {
            expect(args).toMatchSnapshot();
            return response;
        });

        vi.spyOn(response, 'send').mockImplementationOnce((...args) => {
            expect(args).toMatchSnapshot();
            return response;
        });

        errorMiddleware(error, request, response, nextFn);
    });

    it('should send formatted response for AppError error', async () => {
        const request = getRequest();
        const response = getResponse();
        const nextFn = vi.fn();

        const error = new AppError({
            message: 'Some message',
            statusCode: 400,
            errorCode: 'error-code',
            payload: {
                additional: 'payload',
            },
        });

        vi.spyOn(response, 'status').mockImplementationOnce((...args) => {
            expect(args).toMatchSnapshot();
            return response;
        });

        vi.spyOn(response, 'send').mockImplementationOnce((...args) => {
            expect(args).toMatchSnapshot();
            return response;
        });

        errorMiddleware(error, request, response, nextFn);
    });
});
