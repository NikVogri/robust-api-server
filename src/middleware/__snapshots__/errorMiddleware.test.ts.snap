// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test errorMiddleware > should send formatted response for AppError error 1`] = `
[
  400,
]
`;

exports[`Test errorMiddleware > should send formatted response for AppError error 2`] = `
[
  {
    "additional": "payload",
    "errorCode": "error-code",
    "message": "Some message",
  },
]
`;

exports[`Test errorMiddleware > should send formatted response for unexpected error 1`] = `
[
  500,
]
`;

exports[`Test errorMiddleware > should send formatted response for unexpected error 2`] = `
[
  {
    "errorCode": undefined,
    "message": "Something broke!",
  },
]
`;
